### PR TITLE
CUDA memory profile

### PR DESCRIPTION
**Pull request by erhoo82**
# What does this PR do ?

Profile CUDA memory breakdown. The reference is available in the below link.
https://pytorch.org/blog/understanding-gpu-memory-1/

The memory profile will generate a `pickle file` and a user can load the file in the below page.
https://pytorch.org/memory_viz

# Changelog 
- Added the option to profile a range of training steps to profile the memory usage
- Adjust the training step by the number of micro-batches steps per global step.

# Usage
Add the below knobs to the yaml run config.

```yaml
  memory_profile:                                                                      
     enabled: true                                                                      
     start_step: 2                                                                      
     end_step: 5                                                                        
     rank: 0                                                                            
     output_path: <path/to/out_file>
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by akoumpa — APPROVED** [review]
Hi @erhoo82,

I'm going to approve this as is, with one suggestion:

I searched for `hasattr(self, '_memory_profile_enabled')` and found 11 occurrences. Something like `self.has_memory_profile_enabled()` would result in cleaner code, for example, if the `has_memory_profile_enabled` was defined in modelPT.py.
However, that's not a hard requirement as the `hasattr` is widely used in NeMo.

Thank you so much, this is helpful feature to have in NeMo.
